Add chapter on NLP #137

j-lum opened the pull request:
There's quite a bit of content that I had to prune out because it's quite challenging to explain in simple terms. 

Let me know if there's anything that's still hard to follow

j-lum committed:
Add chapter on NLP

damithc reviewed (commented):
Some quick comments.
* The depth and length seem good.
* Add link to the main list of chapters
* Cite source if using images from elsewhere

Go ahead and get some peer reviews.

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+
+![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>
+From the parts-of-speech tags, a tree-structured representation of grammar can be produced.
+This tree is the result of a _constituency parse_, an task of recognizing a sentence and assigning a syntactic structure to it.
```
`an task` -> `a task`

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+![Parts of speech](partsOfSpeech.png "Parts of speech parse") <br/>
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.
+
+![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>
```
Could the blue borders be removed, or surround the entire image?

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.
+
+![Relation tuple](openIE.png "Constituency Parse") <br/>
+From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What the articles that I have written".
```
is the structure in `What the articles that I have written` intentional?

j-lum replied:
Added quotes around the 'I' similar to the previous examples for emphasis.

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+This technique, Open Information Extraction, is utilized by IBM's Watson to defeat human players in the American game show Jeopardy!<sup>[8](#footnote8)</sup>
+
+### Relations 
+Relations is crucial to the understanding of natural languages.
```
`are`

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+Modern input method editors(IMEs) do more than simply translating input to output in the respective language.
+By tackling the syntactic and relational elements of natural language, IMEs provide autocomplete functionality by suggesting words that are related or occur together frequently. 
+
+For example, typing in "*beijing*" in Moon IME, a Chinese IME, will bring up "*olympics*" as a possible suggestion<sup>[12](#footnote12)</sup>.
```
~in~

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+
+#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
+Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>. 
+For example, the two possible input from users regarding a replacement of digital banking token: 
```
`of a`

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+
+![DBS chatbot](dbsChatbot.png "DBS chatbot") <br/>
+
+can be mapped back to the same intent to which a predefined response can be given. 
```
Could move this above the second image

ang-zeyu commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,170 @@
+## What's next
+1. Get an intuition for the English language
+
+    The Language Instinct by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 
```
Missing link here

ang-zeyu reviewed (commented):
Really succinct and easy to follow 👍 
Just presentation / language nits:

j-lum committed:
Prepare for publishing

Fix grammar errors in main body.
Remove top border in PoS image.
Add link in main ToC.

j-lum commented:
@ang-zeyu Thanks for the through review!

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+* [References](#references)
+
+
+## What is NLP
```
I think we're standardising questions to have question marks so it should be `What is NLP?`

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+
+1. Visually explore the themes present in NLP
+    
+    Standford's [CoreNLP](https://corenlp.run/) provides a visual representation of NLP techniques.
```
Stanford's

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+NLP draws from many other fields of science, from formal linguistics to statistics. 
+The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.
+
+Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes.
```
Might be good to state what the three themes are, i.e. "most of them can be generalized to three themes: Syntax, Semantics and Relations, which are discussed in the next section"

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
+Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.
+
+![Parts of speech](partsOfSpeech.png "Parts of speech parse") <br/>
```
Might be good to increase the image size and add a subheader for an image using the 'pic' syntax below.

```xml
<pic src="partsOfSpeech.png" alt="partsOfSpeech" width="70%">

  <sub> _Figure 1. Parts of Speech_ </sub>

</pic>
```

It would also be more aesthetic to centre the image.

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.
+
+![Constituency Parse](constituencyParse.png "Constituency Parse") <br/>
+From the parts-of-speech tags, a tree-structured representation of grammar can be produced.
```
Likewise, can add a subheader here

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+An approach is to represent the meaning of a sentence as a relationship triple consisting of a *Subject*, *Object* and *Relation*.
+
+![Relation tuple](openIE.png "Constituency Parse") <br/>
+From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What the articles that 'I' have written".
```
"What are the articles that 'I' have written?"

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+### Applications of NLP
+
+#### Input methods editors, e.g. Chinese, Japanese and Korean
+Modern input method editors(IMEs) do more than simply translating input to output in the respective language.
```
Might be good to elaborate on what Method Input Editors are, since it's introduced here for the first time

tiuweehan commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,169 @@
+![DBS chatbot](dbsChatbot.png "DBS chatbot") <br/>
+
+## What's next
+1. Get an intuition for the English language
```
Might be good to bold each of the points to make it stand out more

tiuweehan reviewed (approved):
Interesting read!

j-lum committed:
Clean up main text

Add punctuation to question.
Fix spelling mistakes in Stanford.
Center images.
Bold list text.
Change ToC to match contents.
Box ToC for responsive layouts.

j-lum commented:
Something interesting is that the <markdown> tag ignores all the formatting that it is nested in.
e.g 
`<sub><markdown>text</markdown></sub>` renders without inheriting any styles from `<sub>`. Not sure if this is intended behavior.

jamessspanggg commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+# Introduction to Natural Language Processing (NLP)
+Authors: [Lum Ka Fai Jeffry](https://github.com/j-lum)
+
+* [What is NLP?](#what-is-nlp)
```
Can wrap this TOC in a 
```
<box id="article-toc">`
... 
</box>
```
, this will ensure that the TOC is responsive with the screen re-sizing

jamessspanggg commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+NLP draws from many other fields of science, from formal linguistics to statistics. 
+The goal of NLP is to provide new computational capabilities around human language: for example, holding a conversation, summarizing an article, and so on.
+
+Even though the study of NLP covers a diverse range of tasks, most of them can be generalized to three themes: syntax, semantics and relations.
```
perhaps can italicise the syntax, semantics, relations and allow them to be clickable that navigates to their corresponding sections?

jamessspanggg commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+Authors: [Lum Ka Fai Jeffry](https://github.com/j-lum)
+
+* [What is NLP?](#what-is-nlp)
+* [Core problems in NLP](#core-problems-in-nlp)
```
The section Core problems in NLP does not match with the Themes in NLP header

jamessspanggg commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+      <sub><i>Figure 1. Parts of Speech</i></sub>
+    </pic>
+</p>
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[5](#footnote5)</sup>.
```
in the preview, the footnote 5 seems to not work
<img width="197" alt="Screen Shot 2020-03-27 at 12 44 54 PM" src="https://user-images.githubusercontent.com/32864116/77722458-c38caa80-7028-11ea-8049-e16e8fa3d962.png">

jamessspanggg commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+This family of techniques uses special neural network architectures to model sentences.  
+
+#### Automated customer support, e.g. ChatBot, Clare.AI, NanoRep
+Human-curated databases (e.g. **WordNet**) are used together with techniques like **word2vec** to extract actionable words or phrases<sup>[14](#footnote15)</sup>. 
```
mistake of putting 14 instead of 15

jamessspanggg reviewed (approved):
Good read, planning to take NLP next semester so this was a good brief introduction. Some comments:

Tejas2805 commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+
+
+## What is NLP?
+Natural Language Processing (NLP) is the set of methods for making human language accessible to computers.
```
```suggestion
Natural Language Processing (NLP) is the set of methods that makes human language accessible to computers.
```

j-lum replied:
"for making" has the implication that the methods were tailored for the said purpose.
"that makes" loses a bit of that meaning IMO.

Tejas2805 replied:
Ah I got your point. I just feel the sentence is a bit weirdly framed.

Tejas2805 commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+    </pic>
+</p>
+
+## What's next
```
```suggestion
## What's Next?
```

Tejas2805 commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,195 @@
+
+Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.
+
+### Applications of NLP
```
All headings in this section are quite lengthy. Can you prune them?

Tejas2805 reviewed (commented):
We are aiming for consistency across the book by using `titleCase` for the headings. Have left a few comments and suggestions accordingly

jamessspanggg commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+In English, words can be decomposed to _morphemes_, the minimal unit of meaning (e.g. unhappiness to three morphemes un-happy-ness)<sup>[4](#footnote4)</sup>.
+Similarly, sentences can be tagged with word-level _parts-of-speech_ which describe the meaning of each word in the context of that sentence.
+
+<p align="center">
```
can use the md syntax `<center></center>` to center the images

jamessspanggg commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+    </pic>
+</p>
+
+For example, 'word-level' is tagged as an adjective (JJ) and 'sentences' as a plural noun (NNS)<sup>[6](#footnote5)</sup>.
```
should this be `[5]` instead?

jamessspanggg reviewed (commented):
Just some small nits

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+Recent advances in NLP have given rise to useful tools that have become embedded in our daily lives: 
+spam and phishing classification keeps inboxes sane<sup>[1](#footnote1)</sup>;
+automated chatbots lighten the load on customer support staff and provide customers with immediate feedback<sup>[2](#footnote2)</sup>;
+machine translation bridge the gap between cultures<sup>[3](#footnote3)</sup>. 
```
These references feel a bit random. Usually, the reference should point to the source of the fact given in the sentence. If they are simply examples, best to give them specifically as examples.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+## Themes in NLP
+
+### Syntax
+In order to leverage the power of modern machine learning techniques, raw text must be converted to a general-purpose linguistic structure.
```
Abrupt usage of ML here. The relation is not explained above.

j-lum replied:
Rephrased to match the definition given in the intro.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+This tree is the result of a _constituency parse_, a task of recognizing a sentence and assigning a syntactic structure to it.
+From the tree we can see that the complex phrase 'word-level parts-of-speech' is simply a _noun phrase_.
+Incidentally, this is how Microsoft Word checks for grammatical errors<sup>[6](#footnote6)</sup>.
+Any tree that cannot be parsed may be grammatically incorrect or difficult to understand.
```
Should have used an unrelated sentence in the example so as not to add unnecessary complications (never go meta with examples as it always make the reader's job harder). Was this example taken from elsewhere? if so, cite the source. If you designed the example, consider using a different sentence.

j-lum replied:
Changed it to a simpler sentence.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+</p>
+
+From this representation of the sentence, it is possible for a system to answer novel questions like "Who has written articles?" or "What are the articles that 'I' have written".
+This technique, Open Information Extraction, is utilized by IBM's Watson to defeat human players in the American game show Jeopardy!<sup>[8](#footnote8)</sup>
```
_Watson uses this technique to defeat human player_ seems to convey a misleading message. Isn't it _Watson, who is able to defeat human players, uses this technique_? Even then, I'm not sure there is any direct link between the technique and defeating the human player.

j-lum replied:
Changed the example to some other downstream application.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+The word _proposal_, while not stated in the example, implicitly links the two events together.
+While humans are able to draw the connections instinctually, it is not easy to formalize this process computationally.
+
+There are attempts to create semantic _ontologies_ for natural language, groupings of words by their meanings.
```
Unexplained term: ontology

damithc replied:
Explanation of new terms can be given as tooltips, BTW.

j-lum replied:
Added a tooltip.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+
+Word2vec has been used to analyze behaviour on e-commerce sites as human do not tend to browse for random items but purchase items that are related.
+
+### Applications of NLP
```
This heading should be level 2?

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+
+#### Machine translation, e.g. Google Translate
+**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
+This family of techniques uses special neural network architectures to model sentences.  
```
_an_ algorithm?

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+
+#### Machine translation, e.g. Google Translate
+**seq2seq**<sup>[14](#footnote14)</sup>, a algorithm developed by Google, transforms a sequence to another. 
+This family of techniques uses special neural network architectures to model sentences.  
```
Unexplained term: neural network

j-lum replied:
Added a tooltip.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+
+<a name="footnote1">[1]</a>: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6562150/ <br/>
+<a name="footnote2">[2]</a>: https://azure.microsoft.com/en-us/services/bot-service/ <br/>
+<a name="footnote3">[3]</a>: https://github.com/google/seq2seq <br/>
```
Same as 14?

j-lum replied:
Removed.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+1. **Get an intuition for the English language**
+
+    [The Language Instinct](https://stevenpinker.com/publications/language-instinct) by Steven Pinker provides accessible insight about how humans learn language and the basics of formal linguistics. 
+    This is invaluable especially for those who grew up with English as a first language.
```
The reason for this is not obvious.

j-lum replied:
Rephrased the point to be multilingual.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+</p>
+
+## What's Next
+1. **Get an intuition for the English language**
```
So, NLP is tied to English?

j-lum replied:
Rephrased the point to include linguistics in general.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,197 @@
+    
+1. **Read research papers to gain a deeper understanding of the techniques employed**
+
+    Stay up to date with research through portals like [arxiv sanity preserver](https://www.arxiv-sanity.com/search?q=nlp).
```
I'm not sure we should pretend to guide the reader to _self-learn_  NLP. Is that even practical? Perhaps just point the reader to a good online NLP course?

j-lum replied:
Included a reference to CS224n by Stanford.

damithc reviewed (changes requested):
Good start. Some comments added.

j-lum committed:
Implement changes

Explain conversion of natural language to syntactic structures.
Change example of OIE from Watson to MDS.
Add tooltips for most ML and NLP jargon.
Remove duplicate tooltip.
Rephrase motivation behind book recommendation.
Recommend a course on NLP instead of research papers.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,198 @@
+
+#### Machine translation, e.g. Google Translate
+**seq2seq**<sup>[3](#footnote3)</sup>, a algorithm developed by Google, transforms a sequence to another. 
+This family of techniques uses special <tooltip content="artifical neural networks a computer system made up of a number of simple, highly connected processing elements which translate information to some form of desired output"><i>artificial neural network</i></tooltip>   architectures to model sentences.  
```
>artifical neural networks a computer system made up ...

Something missing in that phrase.

damithc commented on contents/naturalLanguage/nl.md:
```diff
@@ -0,0 +1,198 @@
+    
+1. **Visually explore the themes present in NLP**
+    
+    Stanford's [CoreNLP](https://corenlp.run/) provides a visual representation of NLP techniques.
```
I couldn't get this to run though

j-lum replied:
It can be a bit sluggish to load at times depending on the load of the server.

damithc reviewed (changes requested):
Just a cosmetic nit

j-lum committed:
Fix grammar in tooltip
